src/ProgressLogging.jl

```diff
@@ -3,13 +3,168 @@ module ProgressLogging
+    print(io, isempty(progress.name) ? "Progress" : progress.name)
+    if progress.parentid !== ROOTID
+        print(io, " (sub)")
+    end
+    print(io, ": ")
+    if isnan(progress.fraction)
+        print(io, "??%")
+    else
+        print(io, floor(Int, progress.fraction * 100), '%')
+    end
+    return
+end
+
+const PROGRESS_LOGGING_UUID_NS = UUID("1e962757-ea70-431a-b9f6-aadf988dcb7f")
+
+asuuid(id::UUID) = id
+asuuid(id) = uuid5(PROGRESS_LOGGING_UUID_NS, string(id))
+
+asfraction(progress::Nothing) = NaN
+asfraction(progress::AbstractString) = progress == "done" ? NaN : nothing
+asfraction(progress::Real) = convert(Float64, progress)
+
+
+"""
+    ProgressLogging.asprogress(_, name, _, _, id, _, _; progress, ...) :: Union{Progress, Nothing}
+
+Pre-process log record to obtain a [`Progress`](@ref) object if it is
+one of the supported format.  This is mean to be used with the
+`message` positional argument and _all_ keyword arguments passed to
+`Logging.handle_message`.  Example:
+
+```julia
+function Logging.handle_message(logger::MyLogger, args...; kwargs...)
+    progress = ProgressLogging.asprogress(args...; kwargs...)
+    if progress !== nothing
+        return # handle progress log record
+    end
+    # handle normal log record
+end
+```
+"""
+asprogress(_level, progress::Progress, _args...; _...) = progress
+asprogress(_level, name, _module, _group, id, _file, _line; progress = undef, kwargs...) =
+    if progress isa Union{Nothing,Real,AbstractString}
+        _asprogress(name, id; progress = progress, kwargs...)
+    else
+        nothing
+    end
+
+# `parentid` is used from `@logprogress`.
+function _asprogress(name, id, parentid = ROOTID; progress, _...)
+    fraction = asfraction(progress)
+    fraction === nothing && return nothing
+    return Progress(
+        fraction = fraction,
+        name = name,
+        id = asuuid(id),
+        parentid = parentid,
+        done = progress == "done"
+    )
+end
+
+# To pass `Progress` value without breaking progress monitors with the
+# previous `progress` key based specification, we create an abstract
+# string type that has `Progress` attached to it.  This is used as the
+# third argument `message` of `Logging.handle_message`.
+struct ProgressString <: AbstractString
+    progress::Progress
+end
+
+asprogress(_level, str::ProgressString, _args...; _...) = str.progress
+
+Base.string(str::ProgressString) = str.progress.name
+Base.print(io::IO, str::ProgressString) = print(io, string(str))
+Base.convert(::Type{ProgressString}, str::ProgressString) = str
+Base.convert(::Type{T}, str::ProgressString) where {T<:AbstractString} =
+    convert(T, str.progress.name)
+
+function Base.show(io::IO, str::ProgressString)
+    if get(io, :typeinfo, Any) === ProgressString
+        show(io, string(str.progress))
+        return
+    end
+    print(io, @__MODULE__, ".")
+    print(io, "ProgressString(")
+    show(io, str.progress)
+    print(io, ")")
+end
```

Comment:
Do we have to do this? What does it add?

Reply:
> Using `:typeinfo` is to be avoided where possible

My impression is quite different as `:typeinfo` is explicitly mentioned in `IOContext` public API. It's very useful for producing succinct text output when your objects are in containers.

In this case, it was useful for reading progress log records from `Tests.collect_test_logs` by using something like `[r.message for r in logs]`.

But I do agree that it's better to make objects less magical in general. I guess writing tests and do some debugging along the way will help me find a good balance.